gnovm/pkg/gnolang/store.go

```diff
@@ -73,53 +73,276 @@ type Store interface {
 	LogSwitchRealm(rlmpath string) // to mark change of realm boundaries
 	ClearCache()
 	Print()
+}
+
+// TransactionStore is a store where the operations modifying the underlying store's
+// caches are temporarily held in a buffer, and then executed together after
+// executing Write.
+type TransactionStore interface {
+	Store
+
+	// Write commits the current buffered transaction data to the underlying store.
+	// It also clears the current buffer of the transaction.
 	Write()
-	Flush()
 }
 
-// Used to keep track of in-mem objects during tx.
 type defaultStore struct {
-	alloc            *Allocator    // for accounting for cached items
-	pkgGetter        PackageGetter // non-realm packages
-	cacheObjects     map[ObjectID]Object
-	cacheTypes       map[TypeID]Type
-	cacheNodes       map[Location]BlockNode
-	cacheNativeTypes map[reflect.Type]Type // go spec: reflect.Type are comparable
-	baseStore        store.Store           // for objects, types, nodes
-	iavlStore        store.Store           // for escaped object hashes
+	// underlying stores used to keep data
+	baseStore store.Store // for objects, types, nodes
+	iavlStore store.Store // for escaped object hashes
+
+	// transaction-scoped
+	cacheObjects map[ObjectID]Object          // this is a real cache, reset with every transaction.
+	cacheTypes   hashMap[TypeID, Type]        // this re-uses the parent store's.
+	cacheNodes   hashMap[Location, BlockNode] // until BlockNode persistence is implemented, this is an actual store.
+	alloc        *Allocator                   // for accounting for cached items
+
+	// store configuration; cannot be modified in a transaction
+	pkgGetter        PackageGetter         // non-realm packages
+	cacheNativeTypes map[reflect.Type]Type // reflect doc: reflect.Type are comparable
 	pkgInjector      PackageInjector       // for injecting natives
 	nativeStore      NativeStore           // for injecting natives
 	go2gnoStrict     bool                  // if true, native->gno type conversion must be registered.
 
 	// transient
-	opslog  []StoreOp // for debugging and testing.
 	current []string  // for detecting import cycles.
+	opslog  []StoreOp // for debugging and testing.
+}
+
+type hashMap[K comparable, V any] interface {
```

Comment:
Nitpick: why export the interface methods if the interface itself is only implemented within this same package, and not exported?

Reply:
I guess this is philosophical, but my point of view on deciding whether to export or unexport type methods boils down to "if this type were exported, would this method be part of its public API?".

I've seen this help me especially with refactorings (even if you eventually still keep internal, for instance by making an internal/ package). In practice, I see the unexported methods as the "private" methods in OOP - that is to say, mostly to be used by the exported methods of the same type, and not much elsewhere (unless in other functions closely tied to the type).

Reply:
I'm resolving this as this is now a separate package to begin with